### ISSUE TITLE
Incorporate an account status: Active, Vacation, Closed

### ISSUE THREAD

**Issue by uaktags** [OpenThrone/OpenThrone#267]
Along the lines of #194, in order to facilitate that we need to have a form of account status tracking. 

Accounts will have atleast 4 states: active, vacation, closed, idle/stale

For tracking purposes we probably need something like:
vacation start/end date
closed on date
active already has a join date and last_active
idle/stale could probably reference the last active date if greater than a threshold

Additionally, maybe introduce a "reset" state as well
During a reset, instead of resetting a user's stats, which will then also impact all of the historical records for stats and war logs, we could create a new user id, copy the user's display_name, password_hash, and email, then set the new user to active, set the old user to closed/reset, and provide a field to track the new user id.

So if DasTacoMann (userID-2) resets:
- create new userID (let's say 999)
- set uID999's name, hash, and email to uID 2's
- append -retired to display_name and email for uID 2
- set state to reset/closed for uID 2
- set state to active for uID 999

**Comment by NunosDC** [comment]
Likely will need some other states for Banned, Timeout, Suspended or what not.

**Comment by uaktags** [comment]
Good point, 
so something along these lines:
```
enum AccountStatus {
  ACTIVE
  VACATION
  CLOSED
  IDLE
  RESET
  BANNED
  TIMEOUT
  SUSPENDED
}
```
```
status                AccountStatus @default(ACTIVE)
  vacation_start_date   DateTime? // start of the holiday/vacation
  vacation_end_date     DateTime? //end of it
  closed_on_date        DateTime? //date account was closed
  reset_new_user_id     Int? // the new user ID that was created during reset
  previous_user_id      Int?          // For linking to previous accounts after reset
  account_reset_count   Int           @default(0) // increment this everytime we reset for limit enforcements
  banned_until          DateTime?     // timeouts/suspensions are over
  suspension_reason     String?       // Reason for suspension or ban
  ```
  
  so by default you're active, and we can adjust the status based on the available options above, and have those metrics to track for enforcements. Can only reset account X times per year or something of the sort

**Comment by NunosDC** [comment]
Absolutely!!!

**Comment by MarkTraceur** [comment]
Minor database design thing here: Instead of adding a bunch of new columns to the user table, why not create a new table, and add one row that references it? Something like...

```
User
...
  status        AccountStatus?
...

AccountStatus
  type AccountStatusType @default(ACTIVE)
  start_date DateTime // start of this status
  end_date DateTime? // end of this status (not always set at time of creation)
  reason String? // reason for status change, if specified (e.g., in a suspension or ban)
```

Pseudo-code-y, but you get the idea. The reset tracking could still happen on the user table if you wanted, or you could add a UserReset table, something like this:

```
UserReset
  date DateTime @default(now())
  old_user_id Int
  new_user_id Int
```

...then you can do a SELECT on that table to count the number of resets, rather than tracking them in an integer on a database table.

**Comment by uaktags** [comment]
Yea I like that approach much better. It's getting a little congested with my original approach in test.

**Comment by MarkTraceur** [comment]
No, I'm silly, that's not how database relations work. Instead you'd have an index pointing from the AccountStatus entry to the user ID, and a one-to-many relation on the user model. And you might need to do something clever to relate all the old user IDs in the case of a reset when someone is banned (or someone could reset to evade discipline)

**Comment by uaktags** [comment]
Yep I get it. The Account status log may have the same user id listed multiple times, but the most current one should apply. So even if you're on vacation, you can still get banned or have the account closed.

Logic would just be needed on the enforcement of the most correct/latest status.  Id almost want to expand further on things like adding a who performed what action. You placed yourself on vacation, you became IDLE, Taco banned you, etc.

I'm sitting in the office right now on my phone, but definitely brings an interesting way to better track these things. It's just the logic for the final status that would take a little work.

**Comment by uaktags** [comment]
Couple of events to make sure i test against:

When in vacation mode:
- Let the user know they're in vacation mode if they try to login, see if they want to return to Active
- disable turns/gold/citizen rewards
- maybe keep in Attack list but disable spying and attacking against them
- Let others see they're in vacation mode
- Update their public recruit link to disable recruiting and notify of current status

When Banned/Timeout/etc
- Notify that they can't login
- Remove from Attack list
- Notify the public recruit page the account is banned
